### PR TITLE
Get rid of privs comparison

### LINKED ISSUE #234: mysql_user: get rid of privs comparison
##### SUMMARY
(this is a thing to discuss)

We should analyze possibility to get rid of privs comparison entirely and use just a mechanisms like "try except": let's let the server decide which privs are supported and throw the exception itself.

A **solution** can conceptually look like the following:
- Remove priv comparison
- Add exception handling to catch an exception thrown by the DB (maybe there's something currently present)
- The handler (try except + parsing the error for related keywords) will notify the user and exit gracefully (i.e. without ugly tracebacks, etc.)

However this is a breaking change and we'll require a major release in the future.

Possible scenarios:
**Option 1 (Most votes are for this variant)**:
- Implement the above (or another solution) now
- Test it thoroughly with MariaDB and MySQL
- Merge it after 3.0.0 release (within 2 months)
- Notify users via the changelog that such changes will be released in the major release (now it'll be 4.0.0 in 6-8 months or later)
- Release them finally

**Option 2 (not breaking but downside is one more argument)**:
- Implement the solution above without removing the priv comparison (i.e. in parallel)
- Add a new bool argument called, say, `check_privs` with default `yes`)
- If the option is `no` - comparison will be skipped
- Test it with MySQL, MariaDB
- Release it right now and forget - users will be able to switch back and forth if they need

**Option 3**:
- Same as Option 2 but with a warning to users that the `check_privs` will get `no` as its default in 4.0.0
- Plus announcing this via the changelog
- Finally do it in 4.0.0

**Option 4**:
- ...

##### CONTEXT

- Before 2.3.0 all supported privs were hardcoded in the `VALID_PRIVS` frozenset.
- Since 2.3.0 all supported privs are fetched from the server dynamically with the `SHOW PRIVILEGES` command.
- The issue is that some privs were deprecated, replaced and became aliases and they are not present in the `SHOW PRIVILEGES` output but they were present in the `VALID_PRIVS` frozenset.
- Removing `VALID_PRIVS` brought breaking changes.
- Relates to https://github.com/ansible-collections/community.mysql/pull/233 fixed the breaking change - returned `VALID_PRIVS` and now the module merges it in a frozenset with `show_privs` (fetched from server dynamically) and `EXTRA_PRIVS`.

Relates to https://github.com/ansible-collections/community.mysql/pull/233
Relates to https://github.com/ansible-collections/community.mysql/pull/217

### PR DESCRIPTION
**Pull request by rsicart**
##### SUMMARY
Fixes #234 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
user, mysql_user, mysql_role

##### ADDITIONAL INFORMATION
See #234 for more details.

### CODE REVIEW

**Comment by rsicart** [comment]
For the moment I removed all priv comparison code.


There's already some try/catch blocks which catch `mysql_client.Error` errors at `mysql_user` module level. I was thinking to raise a custom exception inside `privileges_grant()` but I'm not sure that we need more detail than this, catching the error at module level seems already enough. What do you think about that?

**Comment by rsicart** [comment]
Also I'm new with `ansible-test`. I think that currently there's only mysql tests. How do you test mariadb with `ansible-test`? Do you have some tips for that?

I tried locally to set `mariadb_install: true` in setup_mysql target defaults, but I have some errors. It seems more complicated than that.

**Comment by Andersson007** [comment]
@rsicart thanks for doing this!
I'm on vacation this week. @Jorge-Rodriguez , @bmalynovytch could you please take a look?

File: tests/integration/targets/test_mysql_user/tasks/test_priv_append.yml [hunk 1]
```diff
@@ -96,6 +96,25 @@
...
+        append_privs: yes
+        state: present
+      check_mode: '{{ enable_check_mode }}'
+      register: result
+      ignore_errors: true
+
+    - name: Assert that there wasn't a change in privileges if check_mode is set to 'no'
+      assert:
+        that:
+          - "result.changed == false"
```
Comment (Andersson007) [inline tests/integration/targets/test_mysql_user/tasks/test_priv_append.yml]:
```suggestion
          - result is failed
```
is this what we expect? Same for the other test task

**Review by Andersson007 — COMMENTED** [review]
@rsicart thanks for working on this!

If we see CI green, it would be great

**Comment by rsicart** [comment]
CI doesn't pass because of python2. I used a `raise ... from ...` which is python3 only syntax. I didn't fix it because perhaps it's time to go on and forget python 2?
This week I'm on holidays, I do not have my computer to fix this pr.

**Comment by Andersson007** [comment]
> CI doesn't pass because of python2. I used a `raise ... from ...` which is python3 only syntax. I didn't fix it because perhaps it's time to go on and forget python 2? This week I'm on holidays, I do not have my computer to fix this pr.

Unfortunately we can't drop Python 2 support (at least in the near future) because this collection is included in Ansible package and it requires support of Python versions that supported by all supported Ansible versions:)
You could use just `module.fail_json()` instead, I think

**Comment by Andersson007** [comment]
We also don't test against MariaDB. It would be really great to add it to our CI (i'd say it's vital to have it). I created an issue recently where help_wanted https://github.com/ansible-collections/community.mysql/issues/238. So at least for now manual testing against MariaDB will be required.

**Comment by Andersson007** [comment]
@rsicart thanks for working on this!
When it's ready for review, please let us know and:
- Eliminate `Draft`
- Add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment)
- Give feedback on my comment about CI:)

Thank you

**Comment by rsicart** [comment]
I agree that having tests for mariadb is vital.

I just tested manually using docker on my computer and different version of mysql (5.6, 5.7, 8.0) and mariadb (10.4, 10.5).

I tried to add `INVALID` privilege to user `toto@%` using mysql client, in order to view error messages returned by the database.

Then I checked what kind of exception is raised by mysql modules (pymysql, mysqldb) in order to catch them, and I wrote integration tests to complete.

But doing that every time is annoying, automated testing for mariadb should be available in CI too!

File: tests/integration/targets/test_mysql_user/tasks/test_priv_append.yml (hunk 1 above)
Comment (rsicart) [reply tests/integration/targets/test_mysql_user/tasks/test_priv_append.yml]: Good catch

**Comment by Jorge-Rodriguez** [comment]
> But doing that every time is annoying, automated testing for mariadb should be available in CI too!

There has been a plan to implement testing against MariaDB for quite some time now, but it is not quite a trivial task as it turns out so we haven't gotten around to it. Recent changes to the requirements for the testing versions have further delayed that task. I do have a mind to have a go at it if there's no progress on that front but I've been quite caught up otherwise. All in all, this is in the backlog.

**Review by Jorge-Rodriguez — APPROVED** [review]

**Comment by Andersson007** [comment]
@rsicart thanks for working on this! Sorry for the later reply, was on PTO. Everything looks good. Thank you for adding the tests!
@Jorge-Rodriguez thanks for reviewing and merging!

**Comment by Andersson007** [comment]
@rsicart @Jorge-Rodriguez @bmalynovytch i didn't notice that this PR marked with `breaking_change`. Actually this shouldn't affect users, right? What do you think if we release this in 3.0.0 (this or next week)?
I wouldn't wait for 4.0.0 as it hasn't been planned at all:)

**Comment by Jorge-Rodriguez** [comment]
I'm good with 3.0.0

**Pull request merged**